src/lua/protobuf.lua

```diff
@@ -0,0 +1,316 @@
+               'id range [%d, %d]'):format(field_id, field_name,
+               reserved_field_id_min, reserved_field_id_max))
+        end
+        local field_def = {
+            type = field_type,
+            name = field_name,
+            id = field_id,
+        }
+        field_by_name[field_name] = field_def
+        field_by_id[field_id] = field_def
+    end
+    return {
+        type = 'message',
+        name = message_name,
+        field_by_name = field_by_name,
+        field_by_id = field_by_id
+    }
+end
+
+-- Create a enum object suitable to pass into
+-- the protobuf.protocol function.
+--
+-- Accepts a name of an enum and an enum definition
+-- in the following format.
+--
+-- enum_def = {
+--     <value_name> = <value_id>,
+--     <...>
+-- }
+local function enum(enum_name, enum_def)
+    local value_by_name = {}
+    local value_by_id = {}
+    for value_name, value_id in pairs(enum_def) do
+        if value_by_id[value_id] ~= nil then
+            error(('Double definition of enum field %q by %d'):format(
+                value_name, value_id))
+        end
+        value_by_name[value_name] = value_id
+        value_by_id[value_id] = value_name
+    end
+    return {
+        type = 'enum',
+        name = enum_name,
+        value_by_name = value_by_name,
+        value_by_id = value_by_id,
+    }
+end
+
+-- Create a protocol object that stores message
+-- data needed for encoding.
+--
+-- Accepts protocol definition using protobuf.message
+-- and protobuf.enum functions as in example.
+--
+-- protocol_def = {
+--     protocol.message(<message_name>, <message_def>),
+--     protocol.enum(<enumc_name>, <enum_def>),
+--     <...>
+-- }
+local function protocol(protocol_def)
+    local return_protocol = {}
+    local declarations = {}
+    for _, def in ipairs(protocol_def) do
+        local field_name = def['name']
+        local field_type = def['type']
+        if declarations[field_name] then
+            error(('Double definition of name %q'):format(field_name))
+        end
+        if field_type == 'message' then
+            for _, val in ipairs(def['field_by_id']) do
+                local val_type = val['type']
+                local standard = protobuf_types[val_type]
+                local declared = declarations[val_type]
+                if val_type == field_name then
+                    error(('Message %q has a field of %q type ' ..
+                        'Recursive definition is not allowed'):format(
+                        field_name, val_type))
+                elseif not standard and not declared then
+                    declarations[val_type] = false
+                end
+            end
+        end
+        declarations[field_name] = true
+        return_protocol[field_name] = def
+    end
+    for field_type, val in pairs(declarations) do
+        if not val then
+            error(('Type %q is not declared'):format(field_type))
+        end
+    end
```

Comment:
I would leave a brief explanation of the check. For example,

> Detects a message or a enum that is used as a field type in one of messages in the protocol, but not defined in the protocol. Allows a type name be defined after its usage.

Reply:
Added a comment message